scipy/integrate/_bvp.py

```diff
@@ -810,6 +812,10 @@ def solve_bvp(fun, bc, x, y, p=None, S=None, fun_jac=None, bc_jac=None,
         mesh interval ``norm(r / (1 + abs(f)) < tol``, where ``norm`` is
         estimated in a root mean squared sense (using a numerical quadrature
         formula). Default is 1e-3.
+    bc_tol : float, optional
+        Desired absolute tolerance for the boundary condition residuals: `bc` 
+        value should satisfy ``abs(bc) < bc_tol`` component-wise. Equals to
```

Comment:
Maybe we should assign `bc_tol` to `1e-3` by default (not to `tol`), because the two condition are rather independent. What do you think?

Reply:
I prefer the present approach of having one parameter to control both tolerances. If the user needs finer-grain control, then two tolerance inputs can be provided.

Reply:
My objection is that (for example) `tol = 1e-8, bc_tol = 1e-8` doesn't have a fundamental meaning. To me it looks as good as `tol = 1e-8, bc_tol = 1e-3 (default)`. And the latter setting is arguably simpler, that's why I advocate for them (not very strongly though).

Which do you think is a better approach after all? I'm fine with either, slightly prefer `bc_tol=1e-3`.

Reply:
I’m pretty strongly in favor of defaulting to `tol`.  This way, the user doesn’t have to guess which tolerance is controlling conversion.  If at first the answer is not precise enough, the user can just change `tol`, the program will iterate a few more times, and return a more precise result.

Otherwise, a typical pattern would be for the user to change `tol`, find that it has no effect on convergence, and then rummage through the documentation to find `bc_tol` as an additional parameter.

Reply:
Maybe I understand it wrong, but it seems to me that boundary condition violation and "residuals" (deviation of derivatives) are distinct. That is we can get a solution with small residuals and large bc violations (what you observed). Considering this, I think that keeping `tol` and `bc_tol` independent is more conceptually sound. 

And again I can point out that `bc_tol` and `tol` have different meaning and units (absolute vs relative). 

I'm not saying that setting `bc_tol = tol` won't work, but I think it is somewhat misleading. Considering your scenario: 
> This way, the user doesn’t have to guess which tolerance is controlling conversion. If at first the answer is not precise enough, the user can just change tol, the program will iterate a few more times, and return a more precise result.

> Otherwise, a typical pattern would be for the user to change tol, find that it has no effect on convergence, and then rummage through the documentation to find bc_tol as an additional parameter.

I think the correct and sensible behavior is to find our why the found solution is inadequate (either boundary conditions are poorly satisfied or the derivatives doesn't match well enough) and then adjust the corresponding parameter.  In other words scipy (as I understand it) should guide its users to good understanding and intelligent choices, as opposed to being used a black box.

-------

I understand your argument, however I'm still in favor of fixed default `bc_tol` (or fully automatic advanced tolerance selection algorithm, which we decided not to implement).

Let's try to ask @rgommers his opinion on this. (I hope it won't be too difficult to understand the point of the discussion, thanks in advance).

Reply:
I haven't really followed this PR so far. The introduction of `bc_tol` is interesting, we have a number of other functions (IIRC optimizers) where bounds can be exceeded in the final result, and there's no tolerance at all to control that. Adding `bc_tol` for that makes sense to me.

Regarding the default, I'm not sure if it matters much if default is a fixed number or if default is None meaning equal to `tol`. Probably the former is slightly cleaner.

> I think the correct and sensible behavior is to find our why the found solution is inadequate (either boundary conditions are poorly satisfied or the derivatives doesn't match well enough

This is much more important I think - as long as the failure clearly states which tolerance is violated and how to control it, the user can quickly fix the problem.

Reply:
I like how the present code guides the user by listing the residuals separately. I just find the ability to control both tolerances with a single parameter to be simpler. It remains easy, but likely very rare, for each tolerance to be specified separately.

Reply:
Let's keep your (author's) version.